### PR TITLE
SubtypesFactory generator

### PR DESCRIPTION
**Pull request by hez2010**
## What does the pull request do?
Use a source generator to replace the old runtime reflection behavior, to make it AOT-compatible and fast.

## What is the current behavior?
We use runtime reflection to gather all `Easing` types.

## What is the updated/expected behavior with this PR?
All `Easing` types will be generated in compilation time, no runtime reflection is required. Note that other similar scenarios can also make use of the `SubtypesFactoryGenerator` for code generation.

## How was the solution implemented (if it's not obvious)?
Implement a source generator for subtypes instance generation.

For the `Easing` case, we put an attribute on the partial method:

```csharp
[SubtypesFactory(typeof(Easing), "Avalonia.Animation.Easings")]
private static partial bool TryCreateEasingInstance(string type, [NotNullWhen(true)] out Easing? instance);
```

then the generated code:

```csharp
using System;
using System.Collections.Generic;

namespace Avalonia.Animation.Easings
{
    partial class Easing
    {
        private static partial bool TryCreateEasingInstance(string type,out Easing? instance)
        {
            var hasMatch = false;
            (hasMatch, instance) = type switch
            {
                "BackEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.BackEaseIn()),
                "BackEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.BackEaseInOut()),
                "BackEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.BackEaseOut()),
                "BounceEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.BounceEaseIn()),
                "BounceEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.BounceEaseInOut()),
                "BounceEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.BounceEaseOut()),
                "CircularEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.CircularEaseIn()),
                "CircularEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.CircularEaseInOut()),
                "CircularEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.CircularEaseOut()),
                "CubicEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.CubicEaseIn()),
                "CubicEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.CubicEaseInOut()),
                "CubicEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.CubicEaseOut()),
                "ElasticEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.ElasticEaseIn()),
                "ElasticEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.ElasticEaseInOut()),
                "ElasticEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.ElasticEaseOut()),
                "ExponentialEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.ExponentialEaseIn()),
                "ExponentialEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.ExponentialEaseInOut()),
                "ExponentialEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.ExponentialEaseOut()),
                "LinearEasing" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.LinearEasing()),
                "QuadraticEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuadraticEaseIn()),
                "QuadraticEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuadraticEaseInOut()),
                "QuadraticEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuadraticEaseOut()),
                "QuarticEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuarticEaseIn()),
                "QuarticEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuarticEaseInOut()),
                "QuarticEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuarticEaseOut()),
                "QuinticEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuinticEaseIn()),
                "QuinticEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuinticEaseInOut()),
                "QuinticEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.QuinticEaseOut()),
                "SineEaseIn" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.SineEaseIn()),
                "SineEaseInOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.SineEaseInOut()),
                "SineEaseOut" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.SineEaseOut()),
                "SplineEasing" => (true, (Avalonia.Animation.Easings.Easing?)new Avalonia.Animation.Easings.SplineEasing()),
                _ => (false, default(Avalonia.Animation.Easings.Easing?))
            };

            return hasMatch;
        }
    }
}
```

## Alternative solution
Write generated code directly without a source generator.

### CODE REVIEW

**Comment by hez2010** [comment]
Now it's ready for review.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019704-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019743-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by hez2010** [comment]
Any thoughts on this PR?
I think "gathering all types under a namespace" is potentially useful not only for `Easing`, so I create this source generator.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019937-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by MarchingCube** [comment]
@hez2010 From my perspective this change looks good. I was curious about perf so I've benchmarked this against current master.

|    Method |     Mean |   Error |   StdDev |  Gen 0 | Allocated |
|---------- |---------:|--------:|---------:|-------:|----------:|
| CreateOld | 178.4 ns | 4.39 ns | 12.88 ns | 0.0880 |     184 B |
| CreateNew | 100.1 ns | 2.25 ns |  6.57 ns | 0.0880 |     184 B |

It is a lot faster despite losing caching in that dictionary (compiler outputs decent code for this large switch, although there is one extra string comparison in there still).

**Comment by maxkatz6** [comment]
As for today it shouldn't be a problem to have source generators in the main repo. I'm fine with it. 
Probably it won't make much sense to generate same code with xaml compiler either.

File: Avalonia.sln [hunk 1]
```diff
@@ -39,6 +39,8 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Shared", "Shared", "{A689DE
 	ProjectSection(SolutionItems) = preProject
 		.editorconfig = .editorconfig
 		src\Shared\ModuleInitializer.cs = src\Shared\ModuleInitializer.cs
+		src\Shared\RawEventGrouping.cs = src\Shared\RawEventGrouping.cs
```
Comment (MarchingCube) [inline Avalonia.sln:42]: Was this file omitted before by mistake? Since it seems to be unrelated to your PR.

File: src/Avalonia.Base/Input/KeyGesture.cs [hunk 2]
```diff
@@ -155,7 +155,11 @@ private static Key ParseKey(string key)
             if (s_keySynonyms.TryGetValue(key.ToLower(), out Key rv))
                 return rv;
 
+#if NETSTANDARD2_0
```
Comment (MarchingCube) [inline src/Avalonia.Base/Input/KeyGesture.cs:158]: Maybe we can add a small helper somewhere to handle such parsing? Just so we don’t need to sprinkle #if everywhere

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019962-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: Avalonia.sln (hunk 1 above)
Comment (hez2010) [reply Avalonia.sln]: I've removed it from sln.

File: src/Avalonia.Base/Input/KeyGesture.cs (hunk 2 above)
Comment (hez2010) [reply src/Avalonia.Base/Input/KeyGesture.cs]: Done.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019966-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: Avalonia.sln [hunk 3]
```diff
@@ -38,7 +38,9 @@ EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Shared", "Shared", "{A689DEF5-D50F-4975-8B72-124C9EB54066}"
 	ProjectSection(SolutionItems) = preProject
 		.editorconfig = .editorconfig
+		src\Shared\EnumHelper.cs = src\Shared\EnumHelper.cs
```
Comment (MarchingCube) [inline Avalonia.sln:41]: I have a last question before approving :)
Can this simply live in `Avalonia.Base` or it is used in projects not referencing that project?

Comment (hez2010) [reply Avalonia.sln:41]: Yes. I've reverted it.

File: src/Shared/EnumParserHelper.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace Avalonia
+{
+    internal class EnumParserHelper
```
Comment (maxkatz6) [inline src/Shared/EnumParserHelper.cs]: Just add it to the Avalonia.Base.
It has internal helpers and it has internal access attribute. Shared files are a bit a pain.

Comment (hez2010) [reply src/Shared/EnumParserHelper.cs]: > Just add it to the Avalonia.Base.

The helper is also referenced in `Avalonia.DesignerSupport`, if we do this we have to make it public.

Comment (hez2010) [reply src/Shared/EnumParserHelper.cs]: > it has internal access attribute.

Oops didn't notice it before. I will change it now.

Comment (maxkatz6) [reply src/Shared/EnumParserHelper.cs]: You can add internal visibility attribute here https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Properties/AssemblyInfo.cs#L31
We manage internal helpers in this way usually
See https://github.com/AvaloniaUI/Avalonia/tree/master/src/Avalonia.Base/Utilities

Comment (hez2010) [reply src/Shared/EnumParserHelper.cs]: Done.

Comment (hez2010) [reply src/Shared/EnumParserHelper.cs]: I didn't do the same for `src\Shared\SourceGeneratorAttributes.cs`, otherwise the source generator will have to reference `Avalonia.Base`.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019972-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Review by MarchingCube — APPROVED** [review]
Looks good, thanks for implementing our feedback!

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019980-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by jkoritzinsky** [comment]
Please update the source generator to use the IIncrementalGenerator interface instead of ISourceGenerator with an ISyntaxReceiver. Syntax receivers have a really bad performance cliff that can completely destroy the IDE experience. Incremental generators (while harder to author initially), have no such issues.

**Comment by hez2010** [comment]
I've updated the source generator to implement `IncrementalGenerator`.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020001-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
